### ISSUE TITLE
Local commands unsupported in fabric >=2?

### ISSUE THREAD

**Issue by jondkelley** [fabric/fabric#1919]
Do you have a local example?

Is local support totally unsupported since version 1 fabric? There's no good docs by example, and the methods / support is gone or something.

```
fab -f fabfile.py  log_local
No idea what 'log_local' is!
```

python
```
from fabric import Connection
from invoke import task, Exit
from invocations.console import confirm

# pip install invocations

@task
def log_local(c):
    """[Run ls -la on localhost]

    Arguments:
        c {[Connection]} -- [Explicit initial context argument, whose value will be a Connection object at runtime.]
    """
    c.local('ls -la')
```

This was following example here https://github.com/fabric/fabric/issues/1747

**Comment by thrownblown** [comment]
i'm wrestling with this too, and this is what i got
`from fabric import task
from fabric import Connection


user = 'ubuntu'
c = Connection(host='ip.ip.ip.ip', user=user)


@task
def pack(local):
    # Creates a tar gzipped archive of the current directory, the name of the
    # archive must be holbertonwebapp.tar.gz and be place in the local dir
    local.run("tar --exclude='*.tar.gz' -cvzf holbertonwebapp.tar.gz .")


@task
def deploy(local):
    # Uploads the archive to the remote server in the directory /tmp/
    # Creates the directory /tmp/holbertonwebapp
    # Untars the holbertonwebapp.tar.gz archive in /tmp/holbertonwebapp
    c.run('mkdir -p /tmp/holbertonwebapp')
    c.put('holbertonwebapp.tar.gz', remote='/tmp/')
    c.run('tar -C /tmp/holbertonwebapp -xzf /tmp/holbertonwebapp.tar.gz')


@task
def clean(local):
    # Deletes the holbertonwebapp.tar.gz on your local machine
    local.sudo('rm ./holbertonwebapp.tar.gz')`

**Comment by thrownblown** [comment]
@jondkelley try
``` fab log_local ```

**Comment by JamesRamm** [comment]
Hey
@jondkelley  @thrownblown 

tldr; use `run` instead of `local` and don't specify the `--hosts` flag

`local` commands are supported but now there is no difference between `run` and `local`.
If you don't specify `--hosts`, then the `Connection` object falls back to being a regular `Context` object from invoke, and `run` works locally

Try this fabfile:

```python

@task
def test(c):
  print(c)
  c.run("echo HELLO")
```

Running `fab test` will print 'HELLO', plus `c`, which will be a `Context` object.

Running `fab --hosts=<host connection string> test` will print 'HELLO' plus `c` which will be a `Connection` object

**Comment by rectalogic** [comment]
> ```
> fab -f fabfile.py  log_local
> No idea what 'log_local' is!
> ```

fabric `tasks.auto_dash_names` is True by default so you need to do `fab -f fabfile.py  log-local` instead of `log_local`

**Comment by exhuma** [comment]
This issue is also something which stumped me when upgrading to Fabric 2. It is very confusing to have an object behaving differently depending on how fabric is called.

In one project I bypassed this issue by just defining an invoke `tasks.py` and then create fabric `Connection` instances inside this file. This way it becomes way more predictable and explicit.

**Comment by christian-intra2net** [comment]
You can ensure `c` is always a fabric `Connection` by giving a default host name to `task()`:
```
@task(host='localhost')
def test(c):
    c.local('hostname')
    c.run('hostname')
```

**Comment by KantiCodes** [comment]
Have any of you tried to wrap Fabric into your own library (just as it's explained [here ](http://docs.pyinvoke.org/en/stable/concepts/library.html#reusing-invoke-s-cli-module-as-a-distinct-binary)for invoke)

**Comment by exhuma** [comment]
@KantiCodes how is this related to this issue?
